server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java

```diff
@@ -141,144 +144,150 @@ static boolean isIdleThread(ThreadInfo threadInfo) {
-                continue;
-            }
-            long cpu = threadBean.getThreadCpuTime(threadId);
-            if (cpu == -1) {
-                threadInfos.remove(threadId);
-                continue;
-            }
-            ThreadInfo info = threadBean.getThreadInfo(threadId, 0);
-            if (info == null) {
-                threadInfos.remove(threadId);
-                continue;
-            }
-            ThreadTimeAccumulator data = threadInfos.get(threadId);
-            if (data != null) {
-                data.setDelta(info, cpu);
-            } else {
-                threadInfos.remove(threadId);
-            }
-        }
-        // sort by delta CPU time on thread.
-        List<ThreadTimeAccumulator> hotties = new ArrayList<>(threadInfos.values());
-        final int busiestThreads = Math.min(this.busiestThreads, hotties.size());
-        // skip that for now
-        final ToLongFunction<ThreadTimeAccumulator> getter = ThreadTimeAccumulator.valueGetterForReportType(type);
-
-        CollectionUtil.introSort(hotties, Comparator.comparingLong(getter).reversed());
-
-        // analyse N stack traces for M busiest threads
-        long[] ids = new long[busiestThreads];
-        for (int i = 0; i < busiestThreads; i++) {
-            ThreadTimeAccumulator info = hotties.get(i);
-            ids[i] = info.info.getThreadId();
-        }
-        ThreadInfo[][] allInfos = new ThreadInfo[threadElementsSnapshotCount][];
+
+        return result;
+    }
+
+    ThreadInfo[][] captureThreadStacks(ThreadMXBean threadBean, long[] threadIds) throws InterruptedException {
+        ThreadInfo[][] result = new ThreadInfo[threadElementsSnapshotCount][];
         for (int j = 0; j < threadElementsSnapshotCount; j++) {
             // NOTE, javadoc of getThreadInfo says: If a thread of the given ID is not alive or does not exist,
             // null will be set in the corresponding element in the returned array. A thread is alive if it has
             // been started and has not yet died.
-            allInfos[j] = threadBean.getThreadInfo(ids, Integer.MAX_VALUE);
+            result[j] = threadBean.getThreadInfo(threadIds, Integer.MAX_VALUE);
             Thread.sleep(threadElementsSnapshotDelay.millis());
         }
-        for (int t = 0; t < busiestThreads; t++) {
-            long time = getter.applyAsLong(hotties.get(t));
-            String threadName = null;
-            for (ThreadInfo[] info : allInfos) {
-                if (info != null && info[t] != null) {
-                    if (ignoreIdleThreads && isIdleThread(info[t])) {
-                        info[t] = null;
-                        continue;
+
+        return result;
+    }
+
+    private void setThreadWaitBlockTimeMonitoringEnabled(ThreadMXBean threadBean, boolean enabled) {
+        if (threadBean.isThreadContentionMonitoringSupported()) {
+            AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
```

Comment:
doPrivileged is required here because we only grant the permission java.lang.management.ManagementPermission "control" to the Elasticsearch core codebase, which would otherwise prevent this to be called with the x-pack plugin. Thanks Ryan!